File: 0542-01-matrix/0542-01-matrix.py
class Solution:
    def updateMatrix(self, mat: List[List[int]]) -> List[List[int]]:
        ROWS, COLS = len(mat), len(mat[0])
        directions = ((1,0), (-1,0), (0,1), (0,-1))
        
        def _inbound(row, col):
            rowInbound = 0 <= row < ROWS
            colInbound = 0 <= col < COLS
            return rowInbound and colInbound
        
        queue = deque()
        
        for r in range(ROWS):
            for c in range(COLS):
                if mat[r][c] == 0:
                    queue.append((r, c))
                else:
                    mat[r][c] = -1      #placeholder until processed
                    
        while queue:
            row, col = queue.popleft()
            
            for dir in directions:
                newRow = row + dir[0]
                newCol = col + dir[1]
                
                if not _inbound(newRow, newCol) or mat[newRow][newCol] != -1:
                    continue
                
                mat[newRow][newCol] = mat[row][col] + 1
                queue.append((newRow, newCol))
                
        return mat

File: 0542-01-matrix/0542-01-matrix.ts
function updateMatrix(mat: number[][]): number[][] {
    // Use BFS approach because we will traverse all neighbors level by level to find shortest distance to zero
    const ROWS = mat.length
    const COLS = mat[0].length
    const result = Array.from({ length: ROWS }, () => Array(COLS).fill(Infinity));
    const queue: [number, number][] = []
    
    // Initialize queue with all 0 and distances
    for (let r=0; r<ROWS; r++) {
        for (let c=0; c<COLS; c++) {
            if (mat[r][c] === 0) {
                result[r][c] = 0
                queue.push([r,c])
            }
        }
    }
    
    // Define directions to move up, down, left, right
    const directions = [
        [-1,0], // up
        [0,1],  // right
        [1,0],  // down
        [0,-1]  // left
    ]
    
    // Do BFS
    while (queue.length > 0) {
        const [r,c] = queue.shift();
        
        for (const [dr, dc] of directions) {
            const newRow = r + dr;
            const newCol = c + dc;
            
            // Check if new pos is in bounds and needs update
            if (newRow >= 0 && newRow < ROWS && newCol >= 0 && newCol < COLS) {
                if (result[newRow][newCol] > result[r][c] + 1) {
                    result[newRow][newCol] = result[r][c] + 1; // this will update distance
                    queue.push([newRow, newCol]) // enqueue to keep going
                }
            }
        }
    }
    
    return result
};